Libplanet/Store/StoreLoaderAttribute.cs

```diff
@@ -0,0 +1,146 @@
+using System;
+using System.Collections.Generic;
+using System.Globalization;
+using System.IO;
+using System.Linq;
+using System.Reflection;
+
+namespace Libplanet.Store
+{
+    /// <summary>
+    /// Registers a static method as a store loader.  The method must comply with
+    /// <see cref="StoreLoader"/> delegate.
+    /// </summary>
+    /// <example>
+    /// <para>With the below code, <c>SampleStore</c> and <c>SampleStateStore</c>-backed
+    /// <see cref="TrieStateStore"/> can be instantiated from a URI like
+    /// <c>sample:///home/foo/bar?cache=512</c>:</para>
+    /// <code><![CDATA[
+    /// [StoreLoader("sample")]
+    /// private static (IStore Store, IStateStore StateStore) Loader(Uri storeUri)
+    /// {
+    ///     NameValueCollection query = HttpUtility.ParseQueryString(storeUri.Query);
+    ///     int cache = query.GetInt32("cache", 128);
+    ///     var store = new SampleStore(storeUri.LocalPath, cache);
+    ///     var stateStore = new SampleStateStore(storeUri.LocalPath, cache);
+    ///     return (store, stateStore);
+    /// }
+    /// ]]></code>
+    /// </example>
+    [AttributeUsage(AttributeTargets.Method, AllowMultiple = true)]
+    public class StoreLoaderAttribute : Attribute
+    {
+        /// <summary>
+        /// Instniates a new <see cref="StoreLoaderAttribute"/> with the specified
+        /// <paramref name="uriScheme"/>.
+        /// </summary>
+        /// <param name="uriScheme">The URI scheme to associate with the attributed store loader
+        /// method.</param>
+        public StoreLoaderAttribute(string uriScheme)
+        {
+            UriScheme = uriScheme.ToLowerInvariant();
+        }
+
+        /// <summary>
+        /// The URI scheme to associate with the attributed store loader method.
+        /// </summary>
+        public string UriScheme { get; }
+
+        /// <summary>
+        /// Loads a pair of <see cref="IStore"/> and <see cref="IStateStore"/> from the specified
+        /// <paramref name="storeUri"/>.
+        /// </summary>
+        /// <param name="storeUri">A URI referring to a store.</param>
+        /// <returns>A pair of loaded <see cref="IStore"/> and <see cref="IStateStore"/>.</returns>
+        public static (IStore Store, IStateStore StateStore)? LoadStore(Uri storeUri)
+        {
+            const BindingFlags flags =
+                BindingFlags.Public |
+                BindingFlags.NonPublic |
```

Comment:
Just curious, why also `NonPublic`?

Reply:
Without this, `internal`/`protected`/`private` methods are not discovered.

Reply:
Yes, I understand that. Is this because due to already implemented access level? Not a big fan of rummaging through `internal` and `private` unless absolutely necessary. 😕

Reply:
I didn't wanted add more unnecessary (these loader methods would never be called directly, as there is no reason to prefer them over regular constructors) public methods, because we should maintain them after once they become public.

Reply:
Well, if the intention is to specifically hide them, then I can get behind NonPublic flag. But in that case, maybe we can get rid of Public? 🤔

Reply:
I indeed considered to get rid of it, but decided to let it discover any methods conforming the proper signature regardless of their visibility, because it would be very difficult to troubleshoot if someone write their own loader method with wrong visibility… We may be able to build a Roslyn analyzer to warn such mistakes in the future.